src/Functional.Unions.FluentAssertions/Functional.Unions.FluentAssertions/UnionValueTypeAssertions.cs

```diff
@@ -0,0 +1,71 @@
+﻿using FluentAssertions;
+using FluentAssertions.Execution;
+using FluentAssertions.Primitives;
+using Functional.Unions.FluentAssertions.Extensions;
+using System;
+using static Functional.Unions.FluentAssertions.Extensions.UnionExtensions;
+
+namespace Functional.Unions.FluentAssertions
+{
+	public class UnionValueTypeAssertions<TUnionType, TUnionDefinition, TOne, TTwo>
+		where TUnionType : struct
+		where TUnionDefinition : UnionDefinitionBase<TUnionType, TUnionDefinition, TOne, TTwo>
+	{
+		private readonly IUnionValue<UnionDefinitionBase<TUnionType, TUnionDefinition, TOne, TTwo>> _subject;
+
+		public UnionValueTypeAssertions(IUnionValue<UnionDefinitionBase<TUnionType, TUnionDefinition, TOne, TTwo>> unionValue)
+		{
+			_subject = unionValue;
+		}
+
+		/// <summary>
+		///  Verifies that the subject is equal to an expected value.
+		/// </summary>
+		/// <param name="expected">The expected value.</param>
+		/// <param name="because">Additional information for if the assertion fails.</param>
+		/// <param name="becauseArgs">Zero or more objects to format using the placeholders in <paramref name="because"/>.</param>
+		/// <returns></returns>
+		public AndConstraint<ObjectAssertions> Be(IUnionValue<UnionDefinitionBase<TUnionType, TUnionDefinition, TOne, TTwo>> expected, string because = "", params object[] becauseArgs)
+		{
+			Execute.Assertion
+				.ForCondition(_subject.Equals(expected))
+				.BecauseOf(because, becauseArgs)
+				.FailWith(() => GetFailReasonForBe<TUnionDefinition>(_subject, expected));
+
+			return new AndConstraint<ObjectAssertions>(new ObjectAssertions(_subject));
+		}
+
+		/// <summary>
+		///  Verifies that the subject's value is of a particular type in the Union
+		/// </summary>
+		/// <typeparam name="TExpected">The type in the Union that the value is expected to be of.</typeparam>
+		/// <param name="because">Additional information for if the assertion fails.</param>
+		/// <param name="becauseArgs">Zero or more objects to format using the placeholders in <paramref name="because"/>.</param>
+		/// <param name="ignore">Ignore this parameter as it should always be the default value.</param>
```

Comment:
I wonder if you could instead have a single `BeOfUnionType<T>` method that does an ugly type-checking chain?  The intent is that the individual types in the union are different, right?

```
public AndUnionValueConstraint<T> BeOfUnionType<T>(...)
{
    if (_subject.One() is Option<T> one)
        return AndUnionValueConstraint<T>(one.ThrowOnFailure(...));
    if (_subject.Two() is Option<T> two)
        return AndUnionValueConstraint<T>(two.ThrowOnFailure(...));

    throw new InvalidOperationException(...);
}
```

Reply:
Yes, that's a good point. The only down-side to this approach is that it is possible for the developer to specify a Type which doesn't exist in the `Union`. I'm not sure if that is of any concern though as it shouldn't be too difficult for the developer to stay on top of which types are a part of each Union. If this is acceptable I can update to your suggestion of a single method with a type checking chain.

I do like the idea of some assertion to ensure that the type are all different. This also raises the question of should we add a method for checking types by position in the `Union` in case the developer has a `Union` with duplicated types?

```C#
public AndUnionValueConstraint<T> BeOfTypeOne<T>(...)
public AndUnionValueConstraint<T> BeOfTypeTwo<T>(...)
...
```

Reply:
Should we just be defining methods `BeOfTypeOne`, `BeOfTypeTwo`, etc?  I think that would be less awkward than having to explicitly specify the type.  Additionally, there is an incentive to not change the types and ordering of discriminated unions due to the breaking changes involved.

```
public class UnionValueTypeAssertions<TUnionType, TUnionDefinition, T1, T2>
{
	public AndUnionValueConstraint<T1> BeOfTypeOne(...)
	{
        	Execute.Assertion
			.BecauseOf(because, becauseArgs)
			.ForCondition(_subject.One().HasValue())
			.FailWith(() => ...);

		return new AndUnionValueConstraint<TExpected>(_subject.One().ThrowOnNone(...));
	}

	public AndUnionValueConstraint<T2> BeOfTypeTwo(...) => ...
}
```

Reply:
This is a good point to think about. I've implemented this methods and given them a bit of a test. 

Implementation-wise they are much cleaner. I definitely like not having to do the ignore params hack or allowing users to enter invalid types.

On the cons side I have noticed that it may be bit difficult to tell which position of the union you are looking for when using a union definition. Hovering over a union using a definition will display `Union<SomeFailure>`, not the actual list of types so you may find yourself trying each position to find the type you are looking for. It is also slightly less clear what type is being asserted at a glance.

That's mostly speculation on my part though, I'm all for going with the method per position route and seeing how that works in practice.

Reply:
Depending on how many assertions you have to write for a particular union in a test suite, it may be worth providing some custom `BeYourUnionPartType` extension method for that union type (a `YourUnionTypeAssertions` class) that delegates to the `BeOfTypeOne` / `BeOfTypeTwo` extension method underneath.  I realize that the point of these extension methods is to help get rid of boilerplate code, but the "specification" nature of `Functional.Unions` means you'll end up having to write custom code to get the syntax / readability result you want if `BeOfTypeOne`, `BeOfTypeTwo` is not what you want the test code to look like.

Without the extra code, Intellisense should help with the type discoverability, no?  Given `union.Should().`, would `BeOfTypeOne` and `BeOfTypeTwo` not show up as selections?  If so, that helps the writability part at least.

Reply:
Good observations, that makes sense to me. The intellisense does offer some assistance in discoverability and additional extensions are always an option.

I'll be removing the old methods and leaving just `BeOfTypeOne / BeOfTypeTwo` before I create assertions for all `Union` sizes. Thanks for the feedback!